<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<HTML>
<HEAD>
   <TITLE>convert features help</TITLE>
<link rel="stylesheet" type="text/css" href = "main.css" media="screen">
</head>
<body class="info">


<center>
<h2><A HREF="RSAT_home.cgi">RSA-tools</A> - convert-features - manual</H2>
</center>

<H4>NAME</H4>
	convert-features
<br/>
        2002 by Jacques van Helden (jacques.van.helden@ulb.ac.be)
	
<H4>DESCRIPTION</H4>

	Interconversions between various formats of feature
	description.

<H4>CATEGORY</H4>
	util

<H4>USAGE</H4>
        convert-features [-i inputfile] [-o outputfile] [-v]

<H4>OPTIONS</H4>
<ul>
<li>	-h	display full help message</li>
<li>	-help	display options</li>
<li>	-v	verbose</li>
<li>	-i inputfile
		if not specified, the standard input is used.
		This allows to place the command within a pipe.</li>
<li>	-o outputfile
		if not specified, the standard output is used.
		This allows to place the command within a pipe.</li>
<li>	-from	input format (Supported: dnapat,ft,gft,gff3,gff)</li>
<li>	-to	output format (Supported: dnapat,ft,gft,fasta,gff3,gff)</li>
</ul>
<H4>INPUT FORMATS</H4>
<ul>
 <li>      <b>ft</b>: RSAT feature-map (extension .ft)</li>

 <li>      <b>dnapat</b>: RSAT dna-pattern (extension .tab)</li>

 <li>      <b>gft</b>: RSAT Genome features (file Feature.tab in the directory data/genomes)</li>

 <li>      <b>gff</b>: General Feature Format as specified by the Sanger insitute (extension .gff)
       	    <a href="http://www.sanger.ac.uk/Software/formats/GFF/GFF_Spec.shtml">http://www.sanger.ac.uk/Software/formats/GFF/GFF_Spec.shtml</a></li>
       
  <li>     <b>gff3</b>: Generic feature format version 3
       	     An adaptation of GFF by Lincoln Stein
       	     <a href="http://flybase.net/annot/gff3.html">http://flybase.net/annot/gff3.html</a></li>
</ul>
<H4>OUTPUT FORMATS</H4>
<ul>
<li>     <b>ft</b>
<li>     <b>gff</b>
<li>     <b>gff3</b>
</ul>
<H4>FORMAT DESCRIPTIONS</H4>

 <p>      Feature formats are generally tab-delimited files. For
       historical reasons, different formats have been used at
       different sites. Originally, these formats were conceived to
       represent different types of information, and, for this reason,
       the contents of their column slightly differs. However, the
       main information is similar, and the differing columns can be
       easily extrapolated or skipped.  </p>
<ul>
       ft: RSAT feature-map
<ul>
<li> 		col 1   map label (eg gene name) </li> 
<li> 		col 2 	feature type</li> 
<li> 		col 3 	feature identifier (ex: GATAbox, Abf1_site)</li> 
<li> 		col 4 	strand (D for Direct, R for Reverse),</li> 
<li> 	col 5 	feature start position</li> 
<li> 		col 6 	feature end position</li> 
<li> 		col 7 	(optional) description </li> 
<li> 		col 8 	(optional) score</li> 
</ul>
       gft: RSAT Genome features (file Feature.tab in the directory data/genomes)
<ul>
<li>		col 1	id</li>
<li>		col 2	type</li>
<li>		col 3	name</li>
<li>		col 4	contig</li>
<li>		col 5	start</li>
<li>		col 6	end</li>
<li>		col 7	strand</li>
<li>		col 8	description</li>
<li>		col 9	chrom_position</li>
<li>		col 10	organism</li>
</ul>
       gff: Sanger general feature file (extension .gff)
           Information: http://www.sanger.ac.uk/Software/formats/GFF/GFF_Spec.shtml
<ul>
<li>		col 1	seqname (contig or sequence ID)</li>
<li>		col 2	source</li>
<li>		col 3	feature (feature type)</li>
<li>		col 4	start</li>
<li>		col 5	end</li>
<li>		col 6	score</li>
<li>		col 7	strand (+, - or .)</li>
<li>		col 8	frame (0, 1, 2 or .)</li>
<li>		col 9	attribute</li>
</ul>
       gff3: Generic Feature Format version 3
           Information: http://flybase.net/annot/gff3.html
<ul>
<li>		col 1	seqid (contig or sequence ID)</li>
<li>		col 2	source</li>
<li>		col 3	feature type</li>
<li>		col 4	start</li>
<li>		col 5	end</li>
<li>		col 6	score</li>
<li>		col 7	strand (+, - or .)</li>
<li>		col 8	frame (=phase) (0, 1 or 2)</li>
<li>		col 9	attributes</li>
<ul>		    	ID:		name of the feature</li>
<li>			Name: 		display name for the feature</li>
<li>			Alias:		secondary name for the feature</li>
<li>			Parent:		parent of the feature</li>
<li>			Target:		target (of an alignment)</li>
<li>			Gap:		alignment of the feature to the target</li>
<li>			Note:		A free text note</li>
<li>			Dbxref:		database cross reference</li>
<li>			Ontology_term:	cross-reference to an ontology term</li>
</ul>
</ul>
</ul>
<HR SIZE=3>

<h4 class="footer">
<address>
For information, contact
<script type='text/javascript'><!--
var v2="6JW4BT3V6VUHNUC3AD4ZJZKSYJC5";var v7=unescape("%7C+4E71@x@7%3Bf%060/W%24*t/%268e2%3Ad%21P");var v5=v2.length;var v1="";for(var v4=0;v4<v5;v4++){v1+=String.fromCharCode(v2.charCodeAt(v4)^v7.charCodeAt(v4));}document.write('<a href="javascript:void(0)" onclick="window.location=\'mail\u0074o\u003a'+v1+'?subject='+'\'">'+'Jacques van Helden<\/a>');
//--></script><noscript><a href='http://w2.syronex.com/jmr/safemailto/#noscript'>Jacques van Helden (using spam protection)</a></noscript>
</address>
</h4>
</body>
</html>
