<HTML>
<HEAD>
   <TITLE>RSA-tools - pattern search manual</TITLE>
</HEAD>
<BODY TEXT="#000000" BGCOLOR="#FFEEDD" LINK="#0000EE" VLINK="#55008B" ALINK="#FF0000">

<H3 ALIGN=CENTER><A HREF="RSA_home.cgi">RSA-tools</A> - pattern search manual</H3>

<H4>Name</H4>
<UL>
dna-pattern<P>
1997-98 by <a HREF="mailto:Jacques van Helden <jvanheld@scmbb.ulb.ac.be>">Jacques van Helden (jvanheld@scmbb.ulb.ac.be)</A>.
</UL>

<H4>Description</H4>
<UL>
Searches all occurrences of a pattern within DNA sequences. The pattern can be entered as a simple nucleotide sequence, or may include degenerate nucleotide codes and regular expressions. 
</UL>

<H4>Options</H4>

<UL>
<P><A NAME<="patterns">
<B>Query pattern(s)</B></A><P>

<UL>
The pattern(s) you are searching. Each pattern should be the first word
of a new line. An identifier can be provided after each pattern, separated
by a tab character. <P>

<B>Example:</B>
<PRE>        TTGTT   TGbox
        GATWA   GATAbox
        CCCCT   STRE</PRE>

<A HREF="http://www.chem.qmw.ac.uk/iupac/misc/naseq.html">Ambiguous
nucleotide codes</A> of the IUPAC-IUB commission are supported. <P>

<PRE>                A                       (Adenine)
                C                       (Cytosine)
                G                       (Guanine)
                T                       (Thymine)
                R       = A or G        (puRines)
                Y       = C or T        (pYrimidines)
                W       = A or T        (Weak hydrogen bonding)
                S       = G or C        (Strong hydrogen bonding)
                M       = A or C        (aMino group at common position)
                K       = G or T        (Keto group at common position)
                H       = A, C or T     (not G)
                B       = G, C or T     (not A)
                V       = G, A, C       (not T)
                D       = G, A or T     (not C)
                N       = G, A, C or T  (aNy)</PRE>

Upper and lower case are considered equivalent. <BR>
The pattern can also contain regular expression elements:<P>

<UL>
<LI><TT>GAT[TA]AG</TT> means &quot;<TT>GATAAG</TT> or <TT>GATTAG</TT>&quot;
(equivalent to <TT>GATWAG</TT>).</LI>

<LI><TT>CGGN{11}CCG</TT> means <TT>CGG </TT>followed by 10 <TT>N</TT> followed
by <TT>CCG</TT>.</LI>

<LI><TT>GATAAGN{0,30}GATAAG</TT> means two <TT>GATAAG </TT>spaced by 0
to 30 nucleotides.</LI>
</UL>
</UL><P>

<P><A NAME="sequence"></A><B>Input sequence</B><P>

<UL>
The sequence in which the pattern has to be searched. </UL>

<P><A NAME="sequence_format"></A>
<B>Input sequence format:</B><BR>
The following sequence formats are supported(click on the links for more details). 
<UL>
<LI><A HREF="help.formats.html#raw"><B>raw</B></A>: the raw sequence without any identifier or comment.
<LI><A HREF="help.formats.html#multi"><B>multi</B></A>: several raw sequences concatenated.
<LI><A HREF="help.formats.html#ig"><B>IG</B></A>: IntelliGenetics format.
<LI><A HREF="help.formats.html#fasta"><B>FastA</B></A>: the sequence format used by FastA, BLAST, Gibbs sampler and a lot of other bioinformatic programs.
<LI><A HREF="help.formats.html#wconsensus"><B>Wconsensus</B></A>: the format defined by Jerry Hertz for his programs (patser, consensus, wconsensus).
</UL><P>

<P><A NAME="strands"></A><B>Search strands</B><P>

<UL>
The pattern can be searched on the direct, 
the reverse complement, 
or both strands.</UL>

<P><A NAME="subst"></A><B>Substitutions</B><P>

<UL>
Imperfect matches can be allowed, with a given number of substitutions.
insertions and deletions are not supported.</UL>

<P><A NAME="return"></A><B>Return</B><P>

<UL>
The program either returns the matching positions, or the number of
matches within each input sequence.</UL>

<P><A NAME="threshold"></A><B>Threshold</B><P>
<ul>
(only valid with the option return count)<br>
Only return sequences having more matches than the specified threshold
</ul>


<P><A NAME="flanking"></A><B>Flanking</B><P>
<UL>
When the option &quot;return matching position&quot; is selected, the
program also return the sequence of the matching word. This sequence can
be extended leftwards and rightwards with the flanking option. 
</UL>


<P><A NAME="origin"></A><B>Origin</B><P>
<UL>
Matching positions can be calculated either relative to the sequence start, or to the sequence end. The second option is most useful for upstream sequences, since it directly indicates the distance between the site and the ORF start. It is thus selected by default. 
</UL>

<P><A NAME="noov"></A><B>Prevent overlapping matches</B><P>
<UL>
This option allows to prevent successiv matches of a periodic pattern to overlap with each other. When the option is activated, each occurrence of a given pattern will prevent another occurrence of the same pattern before the end of the first matching segment. For example, the sequence TATATATATATA could be considered to contain either 4 (with overlap) or 2 (without overlap) occurrences of the pattern TATATA. 
</UL>


<H4>Availability</H4>
<UL>
The program can be used through its web interface at:
<A HREF="http://scmbb.ulb.ac.be/bioinformatics/rsa-tools">http://scmbb.ulb.ac.be/bioinformatics/rsa-tools</A> <BR>
dna-pattern is a perl script running on unix machines (SUN, SGI have been tested). The web interface is a perl-cgi script.
</UL


<CENTER><HR SIZE=3><P></CENTER>

<CENTER><ADDRESS>For information, contact <A HREF="mailto:jvanheld@scmbb.ulb.ac.be">Jacques
van Helden (jvanheld@scmbb.ulb.ac.be)</A></ADDRESS></CENTER>

</BODY>
</HTML>
