<HTML>
<HEAD>
   <TITLE>RSA-tools: gene-info manual</TITLE>
</HEAD>
<BODY TEXT="#000000" BGCOLOR="#FFEEDD" LINK="#0000EE" VLINK="#55008B" ALINK="#FF0000">

<H2 ALIGN=CENTER><A HREF="RSAT_home.cgi">RSA-tools</A> - gene-info manual</H2>

<H3>Name</H3>
<UL>
gene-info<P>
1997-98 by <a HREF="mailto:Jacques van Helden <jvanheld@scmbb.ulb.ac.be>">Jacques van Helden (jvanheld@scmbb.ulb.ac.be)</A>.
</UL>

<H3>Description</H3>

<UL>
Returns genes whose names match a list of query strings. 
<P>
</UL>

<H3>Options</H3>

<UL>

<P>
<A NAME="organism"></A>
<b>Organism</b>
<ul>
A pop-up menu shows the list of organisms supported on this web site. 
</ul>


<A NAME="feattype"></A>
<h4>Feature type</h4>
<ul>
Currently supported: <b>CDS</b>, <b>mRNA</b>.

<P>
Some Genbank flat files contain only CDSs (e.g. bacteria), some also
contain mRNAs (e.g. <i>A.thaliana</i>). When mRNAs are annotated in
Genbank, their coordinates are stored and can be used.

<P>
The advantage of using mRNA is that, if the mRNA is complete (which is
not always the case), the upstream regions aretrieved relative to the
transcription initiation site, rather than the start codon.

<P>
<b>Remarks</b>
<ol>
<li>One gene can be associated to multiple CDSs and to multiple mRNAs.
<li>Many annotated "mRNAs" seem to be actually CDS (e.g. in June 2003,
12,000 out of 27,000 mRNAs from <i>A.thaliana</i> start with ATG).
</ol>


</ul>


<P>
<A NAME="names"></A>
<b>Names</b>
<ul>
A list of strings or substrings to match with gene names.  <br>
The matching is performed in case-insensitive way. <br>
Regular expressions are supported.

</ul>


<P>
<A NAME="full"></A>
<b>Full string matching</b>
<ul>

When checked, the program only returns genes whose full name matches the
query (no substring matching).
</ul>
<P>

</ul>

<H3>Availability</H3>
<UL>
The program can be used through its web interface at:
<A HREF="http://scmbb.ulb.ac.be/bioinformatics/rsa-tools/" target=_blank>
http://scmbb.ulb.ac.be/bioinformatics/rsa-tools/</A> <BR>

gene-info is a perl script running on unix machines. 
The web interface is a perl-cgi script.
</UL>

<HR SIZE=3>

</BODY>
</HTML>
