<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
    <link rel="stylesheet" type="text/css" href = "main.css" media="screen">
  </head>
  <body class="info">



    <center>

      <h2><a href="RSAT_home.cgi">RSA-tools</a>- Useful links</h2>

    </center>




    <a name=genomes></a>
</ul><h3>Genome information</h3><ul>

  <p><li><a target="_blank" href="ftp://ftp.ncbi.nih.gov/genomes/">NCBI
	complete genomes</a>. Bacterial + a few other genomes installed in
      RSAT are parsed from the NCBI complete genome repository.
  </li></p>

  <p><li><a target="_blank"
	    href='http://www.ensembl.org'>Ensembl</a>. Main data
	    source for metazoan genomes installed in RSAT.</p></li>

</ul><h3>Regulatory sequence analysis</h3><ul>

  <a name=databases></a>
  <h4>Databases on transcriptional regulation</h4>

  <ul>


    <li><a target="_blank" href='http://www.pazar.info/'>PAZAR</a> - One
      stop shopping experience for transcription factors and regulatory
      sequence annotations. It is a software framework for the
      construction and maintenance of regulatory sequence data
      annotations; a framework which allows multiple boutique databases to
      function independently within a larger system (or information
      mall). Our goal is to be the public repository for regulatory
      data. </li>

    <li>
      <a href="http://www.oreganno.org/" target="_blank">ORegAnno</a> - An
      open access database for gene regulatory element and polymorphism
      annotation</li>

    <li>
      <a href="http://rulai.cshl.edu/SCPD/" target="_blank">
	SCPD</a> - A promoter database of the yeast <I>Saccharomyces cerevisiae</li>

    <li>
      <a href="http://rulai.cshl.edu/TRED/" target="_blank">TRED - Transcriptional Regulatory Element Database</a></li>

    <li>
      <a target="_blank" href="http://www.yeastract.com/">YeaSTRACT</a> - Yeast Search for
      Transcriptional Regulators And Consensus Tracking.</li>

    <li>
      <a href="http://www.gene-regulation.com/pub/databases.html#transfac" target="_blank">
	TRANSFAC</a> - The Transcription Factor Database</li>

    <li><a target="_blank" href="http://jaspar.cgb.ki.se/cgi-bin/jaspar_db.pl">JASPAR</a> - The high-quality transcription factor binding profile database.</li><

																				   <li><a href="http://www.isbi.net/" target="_blank"> 
																				       OOTFD</a> - object-oriented transcription factors database</li>

    <li>
      <a href="http://www.cifn.unam.mx/Computational_Genomics/regulondb/" target="_blank">
	RegulonDB</a> - A Database on Transcriptional Regulation and Genome Organization

      <br><a target="_blank" href="http://regulondb.ccg.unam.mx/index.html">New interface to RegulonDB</a></li>

    <li>
      <a href="http://prodoric.tu-bs.de/" target="_blank">PRODORIC</a> - prokaryotic database of gene regulation.</li>

    <li>
      <a href="http://arep.med.harvard.edu/dpinteract/" target="_blank">DPinteract (E.coli)</a></li>

    <li>
      <a target="_blank" href="http://elmo.ims.u-tokyo.ac.jp/dbtbs/">DBTBS - a database of Bacillus subtilis promoters and transcription factors</a></li>

    <li>
      <a href="http://corg.molgen.mpg.de/" target="_blank">CORG - a database for COmparative Regulatory Genomics.</a></li>

    <li>
      <a href="http://treg.molgen.mpg.de/cgi-bin/pfm_meme_form.pl"
	 target="_blank">T-Reg Comparator</a> This web application allows you to
      compare a set of position weight matrices (PWM) against the T-Reg
      database. T-Reg is a relational database on transcriptional regulation
      that contains a collection of PWMs.</li>


  </ul>
  <h4>Transcription start sites</H4><ul>

    <li>
      <a href="http://www.epd.isb-sib.ch/" target="_blank">EPD</a> - The Eukaryotic Promoter Database. 

    <li>
      <a href="http://biowulf.bu.edu/zlab/PromoSer/" target="_blank">PromoSer</a> - Human, Mouse and Rat promoter extraction service .

    <li>
      <a target="_blank" href="http://dbtss.hgc.jp/index.html">DBTSS</a> Database of Transcription Start Sites. 

  </ul><h4>Retrieval of regulatory sequences</h4><ul>

    <li><b><a target="_blank" href="http://www.ensembl.org">EnsEMBL</a></b> -
      a software system which produces and maintains automatic annotation on
      selected eukaryotic genomes. </li>

    <li><b><a target="_blank" href="http://www.biomart.org">Biomart</a></b>:
      retrieval of genomic sequences from EnsEMBL - BioMart is a
      query-oriented data management system developed jointly by the
      European Bioinformatics Institute (EBI) and Cold Spring Harbor
      Laboratory (CSHL). </li>

    <a name=pattern_disco></a>
  </ul><h4>motif discovery</H4><ul>

    <li><a target="_blank"
	   href="http://seqmotifs.stanford.edu">BioProspector</a></li>

    <li><a target="_blank"
	   href="http://ai.stanford.edu/~iliu/CompareProspector/">CompareProspector</a>
      - Eukaryotic regulatory element conservation analysis and
      identification using comparative genomics</li>

    <li><a target="_blank"
	   href="http://seqmotifs.stanford.edu">MDscan</a></li>

    <li><a target="_blank"
	   href="http://www.cse.ucsc.edu/~kent/improbizer/improbizer.html">Improbizer</a></li>

    <li><a target="_blank"
	   href="http://www.cbs.dtu.dk/services/DNAarray/ann-spec.php">Ann-spec</a></li>

    <li><a target="_blank"
	   href="http://fluff.cs.columbia.edu:8080/domain/mitra.html">MITRA</a>-
      MIsmatch TRee Algorithm for Motif Discovery: Version 1.0</li>

    <li><a target="_blank" href="http://zlab.bu.edu/glam/">GLAM</a> -
      Gapless Local Alignment of Multiple Sequences</li>

    <li><a href="http://www.esat.kuleuven.ac.be/~dna/BioI/Software.html"
	   target="_blank">INCLUSive</a> - A suite of web-based tools for the
      automatic multistep analysis of microarray data</a>
<ul>

  <li><a
	 href="http://www.esat.kuleuven.ac.be/~thijs/Work/MotifSampler.html"
	 target="_blank">Motif Sampler</a> - A gibbs sampler with
    organism-specific higher order (Markov) models for estimating
    prior frequencies.

  <li><a
	 href="http://www.esat.kuleuven.ac.be/~saerts/software/toucan.php"
	 target="_blank">Toucan</a> - A workbench for regulatory sequence
    analysis on metazoan genomes,</a>

</ul>



<li><a href="http://forkhead.cgb.ki.se/YRSA/"
       target="_blank">YRSA</a> - Yeast Regulatory Sequence Analysis

<li><a
       href="http://abstract.cs.washington.edu/~saurabh/YMFWeb/YMFInput.pl"
       target="_blank">YMF</a> - Finds short motifs in DNA sequences

<li>
  <a href="http://ural.wustl.edu/~jhc1/consensus/html/Html/main.html"
     target="_blank">Consensus server</a>


<li><a href="http://ural.wustl.edu/software.html"
       target="_blank"><b><i>patser</i></b>, <b><i>consensus</i></b></a> and
  other stand-alone applications developed by in Gary Stormo's lab.


<li>
  <a href="http://www.ebi.ac.uk/~brazma/patterns.html"
     target="_blank">Sequence Motif Discovery at EBI</a>

<li>
  <a href="http://meme.sdsc.edu/meme/" target="_blank">MEME</a>

<li>
  <a href="http://www-btls.jst.go.jp/MotifExtraction/"
     target="_blank">YEBIS</a>

<li>
  <a target="_blank" href="http://www.isrec.isb-sib.ch/ssa/">Signal Search
    Analysis Server</a>.  SSA is a software package for the analysis of
  nucleic acid sequence motifs that are postionally correlated with a
  functional site such as a transcription initiation site for
  instance.


</ul>
<h4>Comparing discovered motifs with libraries</h4>

<ul>
  <li><a target="_blank"
	 href='http://meme.nbcr.net/meme4_3_0/cgi-bin/tomtom.cgi'>TomTom</a>. Compare
    your DNA motif against a database of known motifs (e.g., JASPAR,
    TRANSFAC, FlyReg, ...). </li>

  <li><a target="_blank"
	 href='http://treg.molgen.mpg.de/cgi-bin/pfm_meme_form.pl'>T-Reg
      Comparator</a> - compare a set of position weight matrices (PWM)
    against the T-Reg database, a relational database on transcriptional
    regulation that contains a collection of PWMs. </li>

  <li><a target="_blank"
	 href='http://www.benoslab.pitt.edu/stamp/'>STAMP</a> - Alignment,
    Similarity, & Database Matching for DNA Motifs</li>
</ul>

<h4>Pattern search with libraries</H4>
<ul>


  <li>
    <a href="http://transfac.gbf.de/TRANSFAC/programs.html" target="_blank">
      Transfac-associated programs</a> (<a href="http://transfac.gbf.de/cgi-bin/matSearch/matsearch.pl" target="_blank">Matinspector</a>,
    
    <a href="http://transfac.gbf.de/cgi-bin/patSearch/patsearch.pl" target="_blank">
      PatternSearch</a>,
    
    <a href="http://transfac.gbf.de/cgi-bin/saga/saga.pl" target="_blank">
      SaGa</a>).</li>
  
  <li>
    <a href="http://bimas.dcrt.nih.gov/molbio/signal/" target="_blank">
      Web Signal Scan Service</a></li>   
  
  <li>
    <a href="http://bimas.dcrt.nih.gov/molbio/proscan/" target="_blank">
      Web Promoter Scan Service</a></li>

  <li>
    <a href="http://cgsigma.cshl.org/jian/HTML/searchputative.html" target="_blank">SCPD - search putative sites</a> 
    
</ul><h4>Detection of cis-regulatory modules (enhancer regions, composite elements)</h4><ul>

  <li><a target="_blank" href="http://luna.bioc.columbia.edu/Target_Explorer/">Target Explorer</a>


</ul><h4>Promoter prediction</h4><ul>

  <li><a target="_blank" href="http://sdmc.lit.org.sg/promoter/">Dragon genome explorer</a>

  <li><a target="_blank" href="http://www.cbs.dtu.dk/services/Promoter/">Promoter
      2.0</a>. Promoter2.0 predicts transcription start sites of vertebrate
    PolII promoters in DNA sequences

  <li><a target="_blank"
	 href="http://promoter.informatik.uni-erlangen.de/">McPromoter V3</a>.
    The Markov Chain Promoter Prediction Server University of
    Erlangen-Nuremberg (Germany). A statistical tool for the prediction of
    transcription start sites in eukaryotic DNA

  <li><a target="_blank" href="http://www.molbiol.ox.ac.uk/promoterscan.htm">Promoter scan
      II</a>. Eukaryotic Pol II Promoter Prediction

  <li><a target="_blank" href="http://www.fruitfly.org/seq_tools/promoter.html">Neural
      Network Promoter Prediction</a>. Eukaryotes and prokaryotes.

</ul>
<h4>Phylogenetic footprinting</H4>
<ul>

  <li><a target="_blank"
	 href="http://seqmotifs.stanford.edu/">CompareProspector</a></li>

  <li><a target=blank
	 href="http://www.bioinfo.rpi.edu/applications/bayesian/bayes/bayes_align12.pl">
      Bayesian Phylogenetic Footprint Homepage</a>

  <li><a href="http://family.caltech.edu/"
	 target="_blank">FamilyJewels</a>. These tools were designed for
    computing and visualizing comparative analyses of two homologous
    eukaryotic genomic regions, with a focus on finding conserved
    non-coding regions of DNA that may encode regulatory linkages.


  <li><a href="http://bioportal.bic.nus.edu.sg/tres/"
	 target="_blank">TRES</a> - Transcription Regulatory Element Search a
    tool for Comparative Promoter Analysis.

  <li><a href="http://bio.cs.washington.edu/software.html"
	 target="_blank">FootPrinter</a>

  <li><a href="http://www.phylofoot.org/cgi-bin/orthoseq.cgi"
	 target="_blank">Orthoseq</a> - alignment and exon prediction in
    orthologous nucleotide sequences.

  <li>
    <a href="http://www.icsi.berkeley.edu/~roded/creme.html">CREME: A
      Framework for Identifying Cis-Regulatory Modules in Human-Mouse
      Conseved Segments</a>

  <li><a href="http://www.phylofoot.org/consite" target="_blank">ConSite</a>


  <li><a href="http://phylofoot.org/" target="_blank">Tools for phylogenetic footprinting purposes</a>

  <li><a target="_blank" href="http://ural.wustl.edu/~twang/PhyloCon/">PhyloCon</a> - Phylogenetic Consensus for Regulatory Motif Identification


</ul><h4>Predicted elements in whole genomes</h4><ul>


  <li><a target="_blank" href="http://rvista.dcode.org/">rVista</a> -
    identification of evolutionary conserved transcription factor binding
    sites</li>

  <li><a target="_blank" href="http://genome.ucsc.edu/">UCSC</a> HMR
    conserved transcription factor binding sites</li>

</ul><h4>Seqence logos</h4><ul>

  <li><a target="_blank" href="http://weblogo.berkeley.edu/logo.cgi">Weblogo</a> - logo
    drawing from aligned sequences</li>

  <li><a target="_blank"
	 href="http://www.biogenio.com/logo/logo.cgi">GENIO/logo</a> - RNA/DNA &
    Amino Acid Sequence Logos</li>

</ul><h4>Other tools for regulatory  sequence analysis</h4><ul>
  <li>
    <a href="http://www.gsf.de/BIODV/" target="_blank">AG Biodiv</a>
    <ul>
      <li><a href="http://www.gsf.de/BIODV/genomeinspector.html" target="_blank">Genome Inspector</a>
    </ul>



</ul>

</ul><h3>Microarray Data Analysis</h3><ul>

  <li>
    <A href="http://www.transcriptome.ens.fr/ymgv/" target="_blank">
      yMGV</a> - yeast Microarray Global Viewer

  <li>
    <a href="http://ep.ebi.ac.uk/" target="_blank">
      Expression profiler</a> - a set of tools for the analysis and clustering of gene expression and sequence data.

  <li>
    <a href="http://afgc.stanford.edu/" target="_blank">Arabidopsis
      Functional Genomics Consortium (AFGC)</a>

  <li>
    <a href="http://genome-www5.stanford.edu/MicroArray/SMD/"
       target="_blank">Stanford Microarray Database</a>


  <li>
    <a href="http://www.catma.org/" target="_blank">CATMA - A Complete
      Arabidopsis Transcriptome Microarray</a>


</ul><h3>Genome centers</h3><ul>


  <h4>Yeast-specific databases</h4><ul>

    <li><a "href=http://www.bru.ed.ac.uk/~orton/yeti.html"
	   target="_blank">YETI</a> - Yeast Exploration Tool Integrator

    <li>
      <a href="http://genome-www.stanford.edu/Saccharomyces/" target="_blank">
	SGD</a> - Saccharomyces Genome Database

    <li>
      <a href="http://www.mips.biochem.mpg.de/proj/yeast/" target="_blank">
	MIPS</a> - Yeast Genome Project Page


    <li>
      <a href="http://www-genome.wi.mit.edu/personal/manoli/yeasts/" target="_blank">
	Yeast genomes at MIT</a>

    <li>
      <a href="http://genome-www.stanford.edu/fungi/help/yeastsMap.html"
	 >Yeast genome comparison at Stanford</a>

    <li>
      <a href="http://genome.wustl.edu/projects/yeast/">
	Saccharomyces Genome Sequencing at the GSC - WUSTL
      </a>

    <li>
      <a href="http://www.proteome.com/YPDhome.html" target="_blank">
	YPD</a> - Yeast Protein Database)

    <li>
      <a href="http://cgsigma.cshl.org/jian" target="_blank">
	SCPD</a> - A promoter database of the yeast <I>Saccharomyces cerevisiae</I>

    <li>
      <a href="http://alize.ulb.ac.be/YTPdb/" target="_blank">
	YTPdb</a> - The yeast transporter database

    <li>
      <A href="http://www.transcriptome.ens.fr/ymgv/" target="_blank">
	yMGV</a> - yeast Microarray Global Viewer


  </ul><h4>Plasmodium falciparum genome</h4><ul>


    <li>
      <a target="_blank" href="http://plasmodb.org/">PlasmoDB</a> - the
      Plasmodium genome database.

  </ul><h4>Arbidopsis genome</h4><ul>

    <li>
      <a target="_blank" href="http://www.arabidopsis.org">
	The Arabidopsis Information Resource
      </a>

  </ul><h4>Human genome analysis</h4><ul>

    <li>
      <a href="http://genome.ucsc.edu/" target="_blank">Genome Browser</a>
      

    <li>
      <a href="http://www.gsc.riken.go.jp/" target="_blank">RIKEN</a>
      


  </ul>

</ul><h3>Tutorials and courses</h3><ul>

  <li><a target="_blank"
	 href="http://www.hgu.mrc.ac.uk/Users/Philippe.Gautier/tutorial/">
      <b>Comparative genomics and phylogenetic footprints</b></a> (by 
    <a target="_blank"
       href="http://www.hgu.mrc.ac.uk/Users/Philippe.Gautier/">
      Philippe Gautier</a>).

    <br>A practical course on the utilization of comparative genomics
    approaches to detect regulatory elements in mammalian
    promoters. Slides with notes + tutorial + on-line help for each step
    of the tutorial.

</ul>
<hr size=3>
</ul>

</body>
</html>

<!--  
      <li>
	<a href= target="_blank"></a>
	-->
