<html>
<head>
   <meta HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
   <title>RSA-tools - tutorials - dyad-analysis</title>
<link rel="stylesheet" type="text/css" href = "../main.css" media="screen">
</head>
<body class="info">
<blockquote>


<center>
<h2><a href="../RSAT_home.cgi">RSA-tools</a> - 
<a href="tutorials.html">Tutorials</a> -
<a href="../dyad-analysis_form.cgi">dyad-analysis</a>
</h2>
</center>




<h2>Contents</h2>

<p>
<ol>
<li><a href=#intro>Introduction</a>
<li><a href=#example>Motif discovery with a small gene family</a>
<li><a href=#interpret>Interpreting the results</a>
<li><a href=#map>Drawing a feature-map with the discovered patterns</a>
<li><a href=#exercises>Additional exercises</a>
</ol>

<a name=intro></a><h2>Introduction</h2>
<ul>

The analysis of word frequencies gives pretty good results for many
families of co-regulated genes, but fails for a specific class of
transcription factors: the Zn cluster proteins. The reason is these
proteins have two distant points of contact with DNA. Each contact
point imposes a specificity over 3 base pairs, but there is an
intermediate region of fixed width but variable content. The width of
the spacing is transcription factor-specific. This kind of patterns is
not only found in yeast, it is also characteristic for the HTH
proteins in prokaryotes, which also bind spaced pair of
trinucleotides.

<p>
We designed a specific algorithm to extract such motifs:
<b>dyad-analysis</b>. the statistical treatment has been described in
detail in
<ul>
<p><li>van Helden, J., Rios, A. F. & Collado-Vides,
J. (2000). Discovering regulatory elements in non-coding sequences by
analysis of spaced dyads. Nucleic Acids Res. 28(8):1808-18.  <A
HREF="http://www.ncbi.nlm.nih.gov:80/entrez/query.fcgi?cmd=Retrieve&db=PubMed&list_uids=10734201&dopt=Abstract">Pubmed
10734201</a>
<br>
[Free access to
the <a href="http://nar.oxfordjournals.org/cgi/reprint/28/8/1808">PDF
file</a>].
</ul>


</ul>
<a name=example></a><h2>Motif discovery with a small gene family</h2>
<ul>

<p>
We will illustrate the usage of dyad-analysis with a family of genes
which are expressed when galactose is provided in the culture
medium. Starting from the set of upstream sequences for these genes,
we will try to extract over-represented motifs. We will first try to
analyse these sequences with oligo-analysis and see how the program
performs with this kind of motifs. Then we will analyse the same
sequences with dyad-analysis and compare the results.

<ol>
<p><li>

Retrieve upstream sequences from -800 to -1 for the following yeast
genes (as you have seen in the <a
href="tut_retrieve-seq.html">tutorial on sequence retrieval</a>). Make
sure to <b>inactivate</b> the option <b>prevent overlap with neighbour
genes</b>.

<pre>
GAL1
GAL2
GAL3
GAL7
GAL10
GAL80
GCY1
MTH1
PCL10
FUR4
</pre>

<p><li>

Once you have the sequences, click the button labelled
<b>oligonucleotide analysis</b>. Leave all other parameters unchanged
and click <b>GO</b>.

<p>Which result do you obtain ? How many patterns were selected ? What
is their significance ?

<p>
Remember that patterns with a significance index (occ_sig) lower than
1 should not be considered too seriously as putative regulatory
elements. They are likely to appear more or less once per random
sequence set. So, when an analysis does not return any pattern with a
significance > 1, it can be considered as a negative answer.

<p>
With the GAL family, <i>oligo-analysis</i> did not detect any pattern
with a significance higher than 1. The program is thus unable to
identify any really significant motif in the upstream sequernces of
the GAL genes.  This comes from the fact that these genes are
regulated by a Zn-cluster protein, Gal4p, which binds a spaced
dyad. This is precisely the type of patterns for
which <i>dyad-analysis</i> has been designed.

<p><li> Click on the icon <b>back</b> of your browser until you come back
to the page with the upstream sequences you retrieved. This time,
click on the button <b>dyad-analysis</b> in the <b>Next step</b> box.

<p><li> The default parameters are to scan pairs of trinucleotides
spaced by any length between 0 and 20. These parameters are
appropriate when you have no a priori idea of the spacing, since they
will evaluate a good range of possible spacing values.

<p>
Note that the computation time is directly proportional to the spacing
range: when 21 possible spacings are tested (from 0 to 20), the
processing can take a few tens of seconds to a few minuts, depending
on the server load.

<p><li> Leave all other parameters unchanged and click <b>GO</b>. You
will have to wait for the answer a bit longer than for
<i>oligo-analysis</i> (it usually takes 20 seconds for this test
case).


</ol>

</ul>
<a name=interpret></a><h2>Interpreting the results</h2>
<ul>

The results of <i>dyad-analysis</i> are displayed in the same format as those of 
<i>oligo-analysis</i>.  In principle, you should already have
performed the <a href=tut_oligo-analysis.html#interpret>tutorial on
oligo-analysis</a>, and you should thus be able to interpret
the <i>dyad-analysis</i> result page.

<p>
<ol>

<li>How many distinct dyads were analyzed ? (this information appears
in the header at the beginning of the result).

<li>How many dyads are selected as significant ?

<li>What is the highest significance index ?

<li>Look now the result of the pattern-assembly, at the bottom of the
result page. How many patterns are assembled into an alignment ?

<li>How many nucleotides of the consensus are specified (different
from N) ?

</ol>

<p>
The pattern-assmebly returned two alignments, but you can easily see
that these alignments are closely related: they only differ by one
substitution. Actually, the pattern-assembly has an option to allow a
given number of substitutions, but with dyads, allowing one
substitution tends to assemble too mny patterns, so we inactivated it.

</ul>
<a name=exercises></a><h2>Additional exercises</h2>
<ul>
<p>
<ol>
<p><li>Analyze the MET family
(see <a href="tut_retrieve-seq.html">tutorial on sequence
retrieval</a>) with the <i>dyad-analysis</i>, and compare the results
to those previously obtained with <i>oligo-analysis</i>. Discuss the
differences.

<p><li>As a negative control, select some families of random genes
(tool <b>random gene selection</b> in the menu <b>Genomes and
genes</b> of the left frame), and apply <i>dyad-analysis</i> to
discover patterns in their upstream regions. Discuss the result.

<p><li>Until now, we only analyzed "easy" cases, since we used groups of
genes which are all regulated b the same factor. In reality, we will
often use these tools to analyze noisy data sets, like those obtained
from microarray data. The programs are quite robust to noise, and are
still able to detect regulatory patternsm even if the data set
contains some genes which are not ergulated by the same transcriptin
factor as the other ones. 

<p>We tested this with various data sets, but you can experience it
yourself, for example with the clusters of genes expressed at
different stages of the cell cycle (Spellman et al., 1998). The
clusters defined by Spellman and co-workers are available in the
<a target=_blank
href=../data/published_data/dyad_analysis_NAR_2000_data/cell_cycle_gene_clusters/>data
repository</a>. Select some of these clusters, and apply the different
motif discovery appraoches described in these tutorials to detect
putative regulatory elements. You can then compare these elements with
those annotated in <a target=_blank
href="http://rulai.cshl.edu/SCPD/">SCPD</a>, and evaluate if the
predicted motifs correspond to binding sites for factors involved in
the cell cycle.

</ol>


</ul>
<p>
You can now come back to the <a href=tutorials.html>tutorial main
page</a> and follow the next tutorials.

</ul>
<hr width="100%">
<h4 class="footer">
<address>
For suggestions or information request, please contact
<script type='text/javascript'><!--
var v2="6JW4BT3V6VUHNUC3AD4ZJZKSYJC5";var v7=unescape("%7C+4E71@x@7%3Bf%060/W%24*t/%268e2%3Ad%21P");var v5=v2.length;var v1="";for(var v4=0;v4<v5;v4++){v1+=String.fromCharCode(v2.charCodeAt(v4)^v7.charCodeAt(v4));}document.write('<a href="javascript:void(0)" onclick="window.location=\'mail\u0074o\u003a'+v1+'?subject='+'\'">'+'Jacques van Helden<\/a>');
//--></script><noscript><a href='http://w2.syronex.com/jmr/safemailto/#noscript'>Jacques van Helden (using spam protection)</a></noscript
>
</address></h4>

</blockquote>
</body>
</html>
