<html>
<head>
   <title>RSA-tools - tutorials - Random models</title>
</head>
<link rel="stylesheet" type="text/css" href = "../main.css" media="screen">
</head>
<body class="info">
<blockquote>


<center>
<h2><a href="../RSAT_home.cgi">RSA-tools</A> - 
<a href="tutorials.html">Tutorials</a> - 
Random models
</h2>
</center>




<p>The prediction of regulatory elements includes a risk of false
predictions. A good motif discovery program includes some
statistical parameters (E-value, information content, significance
index, ...) to estimate the reliability of each prediction. These
statistical estimators may however be more or less adequate, depending
on the adequation between the underlying statistical model and the
type of biological sequences analyzed. 

<p>A practical way to test the rate of false positive returned by
motif discovery programs is to submit sequences which are <i>not</i>
supposed to be co-regulated. In such case, a good program should be
able to return a negative answer, i.e. no pattern at all. 

<p>This web site offers two tools for generating random sequence sets,
in order to estimate the rates of false positive obtained with
different programs: <i>random-sequence</i> and <i>random-genes</i>.

<ol>

<p><li><a href=../random-seq_form.cgi><b>random sequence</b></a>.  This
program generates random sequences on the basis of various
probabilistic background. One interesting model is to use Markov chain
models, calibrated on the complete set off upstream sequences of an
organism. This means that the generated sequences does not exist as a
biological sequences, but it has the same oligonucleotide composition
as the chosen background. 

<p><li><a href=../random-genes_form.cgi><b>random genes</b></a>. This
program selects a random set of gene identifiers in the selected
genome. The selected genes can then be piped to the program
retrieve-seq, which will retrieve the upstream sequences of the
randomly selected genes. In this case, you will analyze real
biological sequences, contrarily to the program random-sequences. The
analysis of random gene sets is very useful to evaluate the rate of
false positives of motif discovery programs.

</ol>

<p>The analysis of random gene selections gives more realistic (and
more severe) conditions than the randomly generated sequences, which
rely on a quite strong assumption about the model underlying sequence
composition (e.g. Markov chain: which order is the most appropriate ?
Do biological sequences really resemble to the result of a Markov
process ? ...).

<h3>Next tutorials</h3>

<ol>
<li><a href=tut_random-seq.html>Random sequence generation</a>
<li><a href=tut_random-genes.html>Random gene selections</a>
</ol>

<p>
You can now come back to the <a href=tutorials.html>tutorial main
page</a> and follow the next tutorials.

</ul>
<hr width="100%">
<h4 class="footer">
<address>
For suggestions or information request, please contact 
<script type='text/javascript'><!--
var v2="6JW4BT3V6VUHNUC3AD4ZJZKSYJC5";var v7=unescape("%7C+4E71@x@7%3Bf%060/W%24*t/%268e2%3Ad%21P");var v5=v2.length;var v1="";for(var v4=0;v4<v5;v4++){v1+=String.fromCharCode(v2.charCodeAt(v4)^v7.charCodeAt(v4));}document.write('<a href="javascript:void(0)" onclick="window.location=\'mail\u0074o\u003a'+v1+'?subject='+'\'">'+'Jacques van Helden<\/a>');
//--></script><noscript><a href='http://w2.syronex.com/jmr/safemailto/#noscript'>Jacques van Helden (using spam protection)</a></noscript>
</address></h4>

</blockquote>
</body>
</html>
