<html>
<head>
   <meta HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
   <title>RSA-tools - tutorials - genome-scale dna-pattern </title>
<link rel="stylesheet" type="text/css" href = "../main.css" media="screen">
</head>
<body class="info">
<blockquote>


<center>
<h2><a href="../RSAT_home.cgi">RSA-tools</A> - 
<a href="tutorials.html">Tutorials</a> - 
<a href="../genome-scale-dna-pattern_form.cgi">genome-scale dna-pattern</a>
</h2>
</center>



<h2>Introduction</h2>
<ul>

<p>
We are now considering the case where we know the binding specificity
for a given transcription factor, and we are looking for all the genes
potentially regulated by this factor. We will use the Nitrogen
regulation as an example.  Response to nitrogen is mediated by the
so-called GATA-box, whose consensus is GATAAG. An interesting property
is that a single GATA-box is insufficient to affect the
transcription. All genes effectively controlled by these elements
possess at least one group of 2 to 4 closely associated GATA-boxes.

</ul>
<h2>Selecting genes on the basis of pattern counts</h2>
<ul>

<p>We will use <b>dna-pattern</b> to select all genes having 3 or more
occurrences of the GATA box in their 500 bp upstream region.

<p>
<ol>
<li>
On the left frame, click <b>genome-scale dna-pattern (strings)</b>.

<li>In the Sequence retrieval options, make sure that the selected
<b>organism</b> is <b>Saccharomyces cerevisiae</b>, and that <b>Sequence
type</b> is set to upstream.

<li>Specify the sequence limits <b>from</b> -500 <b>to</b> -1.

<li>In the <b>Query pattern(s)</b> box, type

<PRE>GATAAG</PRE>

<li>In the <b>Return </b>option, select <b>match counts</b>.

<li>Change the <b>Threshold </b>to 3.

<li>
Click <b>GO</b>
</ol>

<p>
After a few seconds you will see a list of all the yeast genes having
3 or more occurrences of the GATAAG pattern in their 500 bp upstream
sequence. This list contains some genes of completely unknown
function, some other genes with a weak similarity to other protein,
and some genes of known function (generally, those having a 3-letter +
1-number name).

<p>
Notice that, despite the rudimentary criterion used in this search (3
ocurrences of a single hexanucleotide), most of the genes with known
function are associated to nitrogen metabolism (actually, this is one
of the very rare cases where a single pattern count returns decent
results, see remarks below). The genes with unknown function are good
candidates for testing response to Nitrogen, and might be used for
further experimental characterization.

</ul>
<h2>Links to external databases</h2>
<ul>

<p>
Each ORF identifier appears as a hyperlink, which is
connected to the information on this ORF in the yeast genome
database.

<p>
The right column of the result table links each gene to various
databases. By clicking on a link, you will see the information on that
particular gene in the selected database.

<p>
At the bottom of the page, you should see two butons, which allow to
send the gene selection as a whole as query to KEGG, the pathway
database, or yMGV, the yeast gene expression database. Beware, the
link to yMVG can take a bit of time, because profiles drawing are
generated on the fly, and there is one profile per gene and per
microarrya publication. 

<ol>
<p><li>Send your results to <b>yMGV transcription profiles</b>, and wait
for the answer. Try to detect conditions under which most of the
selected genes are either activated or repressed. Are these conditions
associated to modifications of nitrogen supply in the culture medium ?

<p><li>Come bck to the result page. Send now the results to <b>KEGG
pathway coloring </b>. Are the selected genes associated to some
specific pathway ?
</ol>


</ul>
<h2>Remarks</h2>
<ul>

<p>
Selecting genes on the basis of a single pattern count in their
upstream reagions is a very poor way to predict their regulatory
properties. For this tutorial, we applied a very stringent criterion
(at least 3 GATA boxes in a region restricted to 500bp), which
returned a good selection, but selectivity is at the cost of coverage,
and it is obvious that we missed many of the genes regulated by
GATA-boxes.

<p>
When the pattern is relatively large and degenerated (which is not the
case for GATA boxes), matrix-based pattern genome-scale matching (with
patser) provides better results than string-based pattern
counts. However, even in such a case, there is a tradeoff between
sensitivity and coverage.

<p>
We are currently working on more complex approaches, based on
combinations of motifs, which return more accurate results. These
appproaches ar time costly, and are currently not supported on the web
server, but we will soon dvelop a web interface for them.




</ul>
<p>
You can now come back to the <a href=tutorials.html>tutorial main
page</a> and follow the next tutorials.

</ul>
<hr width="100%">
<h4 class="footer">
<address>
For suggestions or information request, please contact 
<script type='text/javascript'><!--
var v2="6JW4BT3V6VUHNUC3AD4ZJZKSYJC5";var v7=unescape("%7C+4E71@x@7%3Bf%060/W%24*t/%268e2%3Ad%21P");var v5=v2.length;var v1="";for(var v4=0;v4<v5;v4++){v1+=String.fromCharCode(v2.charCodeAt(v4)^v7.charCodeAt(v4));}document.write('<a href="javascript:void(0)" onclick="window.location=\'mail\u0074o\u003a'+v1+'?subject='+'\'">'+'Jacques van Helden<\/a>');
//--></script><noscript><a href='http://w2.syronex.com/jmr/safemailto/#noscript'>Jacques van Helden (using spam protection)</a></noscript>

</address></h4>

</blockquote>
</body>
</html>
