<html>
<head>
   <meta HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=iso-8859-1">
   <title>RSA-tools - tutorials - retrieve-seq</title>
<link rel="stylesheet" type="text/css" href = "../main.css" media="screen">
</head>
<body class="info">
<blockquote>


<center>
<h2><a href="../RSAT_home.cgi">RSA-tools</a> - 
<a href="tutorials.html">Tutorials</a> - 
<a href="../retrieve-seq_form.cgi">retrieve-seq</a>
</h2>
</center>


<br/>
In the left frame, under the title <B>Sequence retrieval</B>, click
on the tool "<B>retrieve sequence</B>". A form appears, allowing you
to select the parameters of sequence retrieval.


<h3>Retrieving sequences for a selected set of genes</h3>
<ul>

We will first retrieve upstream sequences for a set of 5 yeast genes
involved in yeast phosphate metabolism.

<OL>

<p><li> In the pop-up menu "<B>Organism</b>", select <I>Saccharomyces
cerevisiae</I>

<p><li> In the text area under "<B>Genes</B>", type the following list
of genes</LI>

<PRE>
YBR093C
PHO8
pho11
PHO81
Pho84
</PRE>

<!--<p><li>Inactivate the option <b>Prevent overlap with neighbour
genes </b>.-->

<p><li>Leave all other options unchanged and click on the button
<b>GO</b>

</OL>

<p>
After a few seconds, the result is displayed, in the form of a link to
the sequences, followed by a series of Buttons.

<p>The reason why sequences are not directly displayed is that users
usually don't need to see them, since they are retrieved for
subsequent analysis with other tools. This is particularly useful to
avoid transferring large amounts of data over the web when you are
analyzing large sequence sets (e.g. several hundreds of genes).  The
default output mode (in the retrieve-seq form) is thus to maintain the
output on the <i><b>server</b></i>, rather than displaying them on
your browser window.


<P>
Even if this server mode has been used, you still have the possibility
to check the sequences a posteriori. In the result page, click on the
link, in order to see the upstream sequence of the selected
genes. Check the sizes of the sequences obtained. Notice that one
seqence is shorter than the other ones. Try to figure out the reason
for this(if you don't know the answer, don't worry, the answer is
below).

<p>Click on the <b>Back</b> button to come back to the previous result
page. Below the link, you can see a series of buttons, which will
allow you to send the retrieved sequences to the next task (as we will
see in the next tutorials). For the time being, do not click these
buttons.


<p><B>Remarks</B>

<ol>

<p><li>Gene names must be separated by carriage returns. Actually,
only the first word of each line is considered as a query. You can add
additional information on a line, but all the text following the first
word of a line will be ignored.

<p><li>Notice that gene names are case-insensitive.

<p><li>Genes can be specified either by their identifier
(e.g. YFL021W) or by their common name (e.g. GAT1).

<p><li>Synonyms are supported: a gene can be associated to several
names. For example, in <i>S.cerevisiae</i>, the ORF YFL021W is
associated with the names GAT1, MEP80 and NIL1. Multipe names are
quite frequent in yeast; they come from the fact that this gene was
characterized in different laboratories independently, and people
realized only later that they had isolated the same gene. The list of
synonyms of each gene can be obtained with the
tool <b>gene-information</b>.

</ol>


</ul>
<h3>Preventing overlap with neighbour genes</h3>
<ul>

<ol>
<p><li>Come back to the <b>retrieve sequence</b> form. 
<p><li>Perform the same operations as above, but this time, inactivate
the option <b>Prevent overlap with neighbour genes</b>.
<p><li>Compare the size of the sequences with the previous result. 
</ol>

</ul>
<h4>Interpretation of the results</h4>
<ul>

<p>
The default size for yeast upstream sequences is 800bp. The reason for
this choice is that 99% of the regulatory elements found in TRANSFAC
(the Transcription Factor Database) are located between -1 and -800,
relative to the start codon.

<p>
However, the median value for the distance between a yeast gene and
its upstream neighbour is about 450 bp. Thus, with a distance of
800bp, most upstream regions would include a fragment of coding
sequence from the upstream ORF, which might bias subsequent analyses.

<p>
The situation is even worse in bacteria, due to the organization of
genes into operons. In <i>Escherichia coli</i>, 25% of the genes have
an upstream neighbour closer than 50bp. When this neighbour is on the
same strand as the gene, it might indicate that they belong to the
same operon. 

<p>
It is thus usually preferable to prevent overlap with neighbour genes.

<!--
<p>A notable exception is the yeast <i>Saccharomyces
cerevisiae</i>. We evaluated the results obtained when this option is
activated and inactivated respectively, and it is clear that many
patterns escape detection when upstream ORFs are clipped. One possible
reason is that the annotations of <i>Saccharomyces cerevisiae</i>
contain many false ORFs (some auhors estimate that there are ~1000
false ORFs in the original annotation).


<p>
This is particularly well examplified with the yeast PHO genes : when
the option is activated, PHO8 and PHO64 upstream regions are truncated
at 180 and 237 base pairs, respectively. In both cases, the putative
upstream ORF is very short (~300bp), and there is no experimental
evidence for the existence of a gene. In addition, several cis-acting
elements have been shown to be located in the regions corresponding to
these questionable ORFs. When the sequences are clipped, the remaining
Pho4p are not sufficiently over-rerpesented for allowing their
detection in a motif discovery programs.

<p>
The PHO family is an extreme case, and when one disposes of a larger
family, it is less likely for false ORFs to prevent detection of
cis-acting elements. However, even in these cases, we noticed a loss
in sensitivity when the option was activated.


<p>
<b>In summary</b>, we propose the following strategy :
<ul>
<li>When working with bacteria, systematically activate the option
<b>prevent overlap with neighbour genes</b>, to prevent including too
many coding sequences when the genes are located in operons.

<li>With the yeast </i>Saccharomyces cerevisiae</i>, inactivate this
option, to circumvent the weaknesses of gene boundary annotations.

<li>With other eukaryotes, I did not perform any evaluation yet, you
shhould do some trials with some well characterized test cases
(e.g. experimentally characterized regulons).

</ul>
-->

</ul>
<h3>Retrieving sequences for all the genes of a genome</h3>
<ul>

As an illustration fo the retrieval of all sequences, we will now use
the same program to retrieve sequences on the other side of the genes
: downstream sequences.

<OL>

<p><li>Open the <b>retrieve sequence</b> form by clicking
<b>retrieve sequence</b> in the left frame.

<p><li>Make sure that the selected organism is <I>Saccharomyces
cerevisiae</I>

<p><li>Besides the label <B>Genes</B>, select "all".

<p><li>Do not enter anything in the "Gene" box (it would be ignored
anyway).

<p><li>Select "downstream" as <B>Sequence type</B>.

<p><li>In the <B>From</b> box, type "1".

<p><li>In the <B>to</b> box, type "50".

<p><li>In the pop-up menu <B>Sequence label</B>, select "gene identifier + name".

<p><li>Click <B>GO</B>

</OL>

<p>
With the above parameters, you retrieved the 50bp located downstream
of the stop codon, for each yeast gene. This sequence sets thus mostly
contains the 3' untranslated region (UTR). In yeast, the 3' UTR is
involved in various functions such as termination of transcription,
RNA maturation (cleavage and poly-adenylation), mRNA stability, and
translational efficiency. The analysis of these sequences allows to
detect some signals involved in such functions (see <a
href=http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=PubMed&list_uids=10648794&dopt=Abstract>van
Helden, J., Olmo, M. & Perez-Ortin, J. E. (2000)</a> for an
application).

</ul>
<h3>Reference position</H3>
<ul>

<p>
The reference position (coordinate 0) depends on the type of sequences
to be retrieved:
<p>

<ul>
<p>
<li>For <B>upstream sequences</B>:

<ul><pre>
     UPSTREAM         ORF
                    +--------> 
                    |start
--------------------------------
...  T  C  A  A  G  A  T  G ...
    -5 -4 -3 -2 -1  0  1  2
</pre></ul>

<ul>
<li>The reference position (coordinate 0) is the ORF start, i.e. the
first nucleotide of the start codon.

<li>The coordinate 0 corresponds to the first nucleotide of the coding
sequences, i.e. the first letter of the start codon.

<li>Negative coordinates are used to indicate sequences located
upstream the start codon, i.e. in the 5'end flank of the gene
(e.g. from -1 to -800 for the yeast).

<li>Negative coordinates are indicate 5' non-coding flank of the gene,
and positive coordinates the 5' side of the coding sequence.

</ul>



<p>
<li>On the contrary, when you select <B>downstream sequences</B>: 

<ul><pre>
      ORF        DOWNSTREAM
  ...------>
        stop
--------------------------------
...  T  A  G  A  C  G  T ...
    -2 -1  0  1  2  3  4
</pre>

<li>The reference point (coordinate 0) is the ORF end, i.e. the last
letter of the stop codon.

<li>Positive coordinates are used to specify sequences located
downstream the stop codon, i.e. the (non-coding) 3'end flank of the gene.

<li>Negative coordinates can be specified to retrieve the (coding) 3'
side of the ORF.


</ul>


</ul>

</ul>
<h3>Exercises</H3>
<ul>

<ol>
<li>On the basis of the above information, extract all start codons
for a given organism (say <i>Saccharomyces cerevisiae</i>). Check that
the codons correspond to the expectation (ATG).

<p>
<li>Extract all the stop codons for the same organism. Check that the
stop codons correspond to the expectation (TAA, TAG or TGA).  </ol>

</ul>
<p>
You can now come back to the <a href=tutorials.html>tutorial main
page</a> and follow the next tutorials.

</ul>
<hr width="100%">
<h4 class="footer">
<address>
For suggestions or information request, please contact 
<script type='text/javascript'><!--
var v2="6JW4BT3V6VUHNUC3AD4ZJZKSYJC5";var v7=unescape("%7C+4E71@x@7%3Bf%060/W%24*t/%268e2%3Ad%21P");var v5=v2.length;var v1="";for(var v4=0;v4<v5;v4++){v1+=String.fromCharCode(v2.charCodeAt(v4)^v7.charCodeAt(v4));}document.write('<a href="javascript:void(0)" onclick="window.location=\'mail\u0074o\u003a'+v1+'?subject='+'\'">'+'Jacques van Helden<\/a>');
//--></script><noscript><a href='http://w2.syronex.com/jmr/safemailto/#noscript'>Jacques van Helden (using spam protection)</a></noscript>

</address></h4>

</blockquote>
</body>
</html>
